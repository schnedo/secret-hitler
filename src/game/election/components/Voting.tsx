import type { ReactElement } from "react";
import type { Phase } from "../..";
import type { PlayerId } from "../../player";
import type { PlayerVotes } from "../index";
import type Vote from "../Vote";

export interface VotingProps {
  playerId: PlayerId;
  onVote: (vote: Vote) => void;
  phase: Phase | null;
  playerVotes: PlayerVotes;
}

export default function Voting({
  playerId,
  onVote,
  phase,
  playerVotes,
}: VotingProps): ReactElement {
  if (phase !== "vote") {
    return <></>;
  }

  if (playerVotes[playerId] !== undefined) {
    return <div>{playerVotes[playerId] ? "Yes" : "No"}</div>;
  }
  return (
    <div>
      <button onClick={() => onVote({ playerId, agreed: true })}>Yes</button>
      <button onClick={() => onVote({ playerId, agreed: false })}>No</button>
    </div>
  );
}
