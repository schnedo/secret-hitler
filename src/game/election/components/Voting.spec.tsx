import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type Vote from "../Vote";
import Voting from "./Voting";

it("should render correctly", async () => {
  expect.hasAssertions();

  const { container } = render(
    <Voting playerId={0} onVote={() => undefined} playerVotes={[]} />,
  );

  expect(container).toMatchSnapshot();
});

it("should call onVote with correct Vote", async () => {
  expect.hasAssertions();

  const playerId = 0;
  const handleVote = jest.fn();
  render(<Voting playerId={playerId} onVote={handleVote} playerVotes={[]} />);
  expect(handleVote).not.toHaveBeenCalled();

  await userEvent.click(screen.getByRole("button", { name: "Yes" }));
  expect(handleVote).toHaveBeenCalledTimes(1);
  let expectedVote: Vote = {
    playerId,
    agreed: true,
  };
  expect(handleVote).toHaveBeenLastCalledWith(expectedVote);

  await userEvent.click(screen.getByRole("button", { name: "No" }));
  expect(handleVote).toHaveBeenCalledTimes(2);
  expectedVote = {
    playerId,
    agreed: false,
  };
  expect(handleVote).toHaveBeenLastCalledWith(expectedVote);
});

it("should show the vote of the player if he has already voted", async () => {
  expect.hasAssertions();

  const playerId = 0;
  const { rerender } = render(
    <Voting
      playerId={playerId}
      onVote={() => undefined}
      playerVotes={[true]}
    />,
  );
  expect(screen.getByText("Yes")).toBeInTheDocument();
  expect(screen.queryByText("No")).not.toBeInTheDocument();

  rerender(
    <Voting
      playerId={playerId}
      onVote={() => undefined}
      playerVotes={[false]}
    />,
  );
  expect(screen.queryByText("Yes")).not.toBeInTheDocument();
  expect(screen.getByText("No")).toBeInTheDocument();
});
