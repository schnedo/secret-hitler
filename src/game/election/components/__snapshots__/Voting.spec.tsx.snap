// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div>
    <button>
      Yes
    </button>
    <button>
      No
    </button>
  </div>
</div>
`;
